### PR TITLE
CLEO testing library

### PR DESCRIPTION
**Pull request by x87**
Goals: 
* Focus on important things when writing unit tests. 
* Standardize test structure. 

Public API:

- `test(name: string, callback: int)`  -- creates new test suite (collection of unit tests)
- `it(expectation: string, callback: int)` -- creates new individual test focusing on testing one thing
- `before_each(callback: int)` -- registers a callback that runs before each unit test (test setup)
- `after_each(callback: int)` -- registers a callback that runs after each unit test (test teardown)
- `assert_eq(val1: int, val2: int)` - checks if two int values are equal, otherwise stops the test execution
- `assert_eqf(val1: int, val2: int)` - same as above, but for floats
- `assert_neq(val1: int, val2: int)` - checks if two int values are not equal, otherwise stops the test execution
- `assert_neqf(val1: int, val2: int)` - same as above, but for floats
- `assert(val: int)` - checks if value is not 0, otherwise stops the test execution
- `assert_ptr(val: int)` - checks if value is a valid pointer, otherwise stops the test execution

Simple example:

```
{$CLEO}
test('test suite', @tests)
terminate_this_custom_script

function tests
  it('should be true', @test1)
  it('should be false', @test2)
  return true

  function test1
   assert_eq(true, true)
  end
  function test2
   assert_eq(false, false)
  end
end
```

### CODE REVIEW

**Comment by MiranDMC** [comment]
`it(` looks bad, almost like `if`
test -> test_suite/block/group
it -> test_case

**Comment by MiranDMC** [comment]
Please port some of file access opcodes. It will show use case of before_each (and lack of after_each?) functionality. At this moment I have no idea how the file handle is meant to be passed into testcase.

**Comment by x87** [comment]
> Please port some of file access opcodes. It will show use case of before_each (and lack of after_each?) functionality. At this moment I have no idea how the file handle is meant to be passed into testcase.

added after_each, see example in 0AC8 (allocate/free)

**Comment by MiranDMC** [comment]
> what C++ libraries are used for unit testing? I can look into how they name things.

I used google tests. Visual Studio also has some built in support.

File: tests/cleo_tests/MemoryOperations/0A8C.txt [hunk 1]
```diff
@@ -1,180 +1,88 @@
...
+        assert_eq(1@, 0xcccccccc)
+        assert_eq(2@, 0xdddddd44)
+        assert_eq(3@, 0xeeeeeeee)
+    end
+
+    function test3
+        write_memory {address} 0@ {size} 2 {value} 0x11223344 {vp} false
+        assert_eq(1@, 0xcccccccc)
+        assert_eq(2@, 0xdddd3344)
+        assert_eq(3@, 0xeeeeeeee)
```
Comment (MiranDMC) [inline tests/cleo_tests/MemoryOperations/0A8C.txt:48]: Assert index would be nice too, as in current error message assert source can only be guessed by expected data (which might be identical for few asserts).

Comment (x87) [reply tests/cleo_tests/MemoryOperations/0A8C.txt:48]: yes, would be nice to point out the assert location.

Comment (x87) [reply tests/cleo_tests/MemoryOperations/0A8C.txt:48]: not sure about explicit index, can be annoying manually tag each assert

Comment (MiranDMC) [reply tests/cleo_tests/MemoryOperations/0A8C.txt:48]: I mean counted in background. I think you did same thing for test indexing.

Comment (x87) [reply tests/cleo_tests/MemoryOperations/0A8C.txt:48]: added index

**Comment by MiranDMC** [comment]
How does the test output looks on the screen?
In my older tests I had multiple `wait 0` commands between test actions to assure in case of game crash the info (of previous step) had chance to be written in to log.

**Comment by x87** [comment]
image

**Comment by MiranDMC** [comment]
> image

Nice. I have finally to change debug utils plugin so it prints top-down, as current order is confusing.

**Review by MiranDMC — APPROVED** [review]

**Comment by x87** [comment]
> Nice. I have finally to change debug utils plugin so it prints top-down, as current order is confusing.

it looks fine to me, especially the fading away effect makes more sense in bottom-top direction, without shifting text back to top

**Pull request merged**